pkg/ceph/mgr/daemon.go

```diff
@@ -0,0 +1,135 @@
+
+import (
+	"fmt"
+	"path"
+	"regexp"
+
+	"github.com/coreos/pkg/capnslog"
+	"github.com/rook/rook/pkg/ceph/mon"
+	"github.com/rook/rook/pkg/clusterd"
+	"github.com/rook/rook/pkg/util"
+	"github.com/rook/rook/pkg/util/proc"
+)
+
+var (
+	logger          = capnslog.NewPackageLogger("github.com/rook/rook", "cephmgr")
+	keyringTemplate = `
+[mgr.%s]
+	key = %s
+	caps mon = "allow *"
+`
+)
+
+const (
+	cephmgr = "ceph-mgr"
+)
+
+type Config struct {
+	InProc      bool
+	ClusterInfo *mon.ClusterInfo
+	Name        string
+	Keyring     string
+}
+
+func Run(context *clusterd.Context, config *Config) error {
+	logger.Infof("Starting MGR %s with keyring %s", config.Name, config.Keyring)
+	err := generateConfigFiles(context, config)
+	if err != nil {
+		return fmt.Errorf("failed to generate mgr config files. %+v", err)
+	}
+
+	_, err = startMgr(context, config)
+	if err != nil {
+		return fmt.Errorf("failed to run mgr. %+v", err)
+	}
+
+	return err
+}
+
+func generateConfigFiles(context *clusterd.Context, config *Config) error {
+	// write the latest config to the config dir
+	/*if err := mon.GenerateAdminConnectionConfig(context, config.ClusterInfo); err != nil {
+		return fmt.Errorf("failed to write connection config. %+v", err)
+	}*/
+
+	keyringPath := getMgrKeyringPath(context.ConfigDir, config.Name)
+	confDir := getMgrConfDir(context.ConfigDir, config.Name)
+	username := fmt.Sprintf("mgr.%s", config.Name)
+	settings := map[string]string{
+		"mgr data": confDir,
+	}
+	logger.Infof("Conf files: dir=%s keyring=%s", confDir, keyringPath)
+	_, err := mon.GenerateConfigFile(context, config.ClusterInfo, confDir,
+		username, keyringPath, false, nil, settings)
+	if err != nil {
+		return fmt.Errorf("failed to create config file. %+v", err)
+	}
+
+	keyringEval := func(key string) string {
+		return fmt.Sprintf(keyringTemplate, config.Name, key)
+	}
+
+	err = mon.WriteKeyring(keyringPath, config.Keyring, keyringEval)
+	if err != nil {
+		return fmt.Errorf("failed to create mds keyring. %+v", err)
+	}
+
+	return nil
+}
+
+func startMgr(context *clusterd.Context, config *Config) (mgrProc *proc.MonitoredProc, err error) {
+
+	// start the mgr daemon in the foreground with the given config
+	logger.Infof("starting ceph-mgr")
+
+	confFile := getMgrConfFilePath(context.ConfigDir, config.Name, config.ClusterInfo.Name)
+	util.WriteFileToLog(logger, confFile)
+
+	keyringPath := getMgrKeyringPath(context.ConfigDir, config.Name)
+	util.WriteFileToLog(logger, keyringPath)
+	args := []string{
+		"--foreground",
+		fmt.Sprintf("--cluster=%s", config.ClusterInfo.Name),
+		fmt.Sprintf("--conf=%s", confFile),
+		fmt.Sprintf("--keyring=%s", keyringPath),
+		"-i", config.Name,
+	}
+
+	if config.InProc {
+		err = context.ProcMan.Run(cephmgr, cephmgr, args...)
+	} else {
+		mgrProc, err = context.ProcMan.Start(cephmgr, cephmgr, regexp.QuoteMeta(cephmgr), proc.ReuseExisting, args...)
+	}
+	if err != nil {
+		err = fmt.Errorf("failed to start mgr: %+v", err)
+	}
+	return
```

Comment:
Shouldnt you return the error if any and also return a *proc.MonitoredProc? Im surprised this compiles since it is not returning anything.